### PR TITLE
Run2-alca91 Modify IsoTrack analysis code to make use of L1 Trigger information

### PR DESCRIPTION
**Pull request by bsunanda**
Can be used to estimate HLT efficiency

### CODE REVIEW

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@lpernie @arunhep @franzoni Can you take a look and approve this?

**Comment by lpernie** [comment]
+1

File: Calibration/HcalCalibAlgos/test/HcalIsoTrkAnalyzer.cc [hunk 1]
```diff
@@ -396,21 +449,75 @@ void HcalIsoTrkAnalyzer::analyze(edm::Event const& iEvent, edm::EventSetup const
...
-			       << ignoreTrigger_;
+  edm::LogVerbatim("HcalIsoTrack") << "# of propagated tracks " << t_TracksProp
+				   << " out of " << t_Tracks << " with Trigger "
+				   << ignoreTrigger_;
 #endif
+
   //Trigger
-  t_trgbits->clear();
-  t_TracksSaved = 0;
+  t_trgbits->clear(); t_hltbits->clear();
```
Comment (davidlange6) [inline Calibration/HcalCalibAlgos/test/HcalIsoTrkAnalyzer.cc]: hi @bsunanda - the manipulations here and just below can be replaced by

t_trgbits->assign(trigNames_.size(),false)

File: Calibration/HcalCalibAlgos/test/HcalIsoTrkAnalyzer.cc [hunk 2]
```diff
@@ -396,21 +449,75 @@ void HcalIsoTrkAnalyzer::analyze(edm::Event const& iEvent, edm::EventSetup const
...
   t_TrigPass    = false;
-  for (unsigned int i=0; i<trigNames_.size(); ++i) t_trgbits->push_back(false);
+  for (unsigned int i=0; i<trigNames_.size(); ++i) {
+    t_trgbits->emplace_back(false);
+    t_hltbits->emplace_back(false);
+  }
+
+  //L1
+  l1GtUtils_->retrieveL1(iEvent,iSetup,tok_alg_);
+  const std::vector<std::pair<std::string, bool> > finalDecisions = l1GtUtils_->decisionsFinal();
```
Comment (davidlange6) [inline Calibration/HcalCalibAlgos/test/HcalIsoTrkAnalyzer.cc]: should be a const &

File: Calibration/HcalCalibAlgos/test/HcalIsoTrkAnalyzer.cc [hunk 3]
```diff
@@ -396,21 +449,75 @@ void HcalIsoTrkAnalyzer::analyze(edm::Event const& iEvent, edm::EventSetup const
...
+  for (auto decision : finalDecisions) {
```
Comment (davidlange6) [inline Calibration/HcalCalibAlgos/test/HcalIsoTrkAnalyzer.cc]: const & decision

File: Calibration/HcalCalibAlgos/test/HcalIsoTrkAnalyzer.cc [hunk 4]
```diff
@@ -396,21 +449,75 @@ void HcalIsoTrkAnalyzer::analyze(edm::Event const& iEvent, edm::EventSetup const
...
+				   << " is " << t_L1Bit << " from a list of "
+				   << finalDecisions.size() << " decisions";
+#endif
+  
+  //HLT
+  edm::Handle<edm::TriggerResults> triggerResults;
+  iEvent.getByToken(tok_trigRes_, triggerResults);
+  if (triggerResults.isValid()) {
+    const edm::TriggerNames & triggerNames = iEvent.triggerNames(*triggerResults);
+    const std::vector<std::string> & triggerNames_ = triggerNames.triggerNames();
```
Comment (davidlange6) [inline Calibration/HcalCalibAlgos/test/HcalIsoTrkAnalyzer.cc]: trailing _ are for class member variables

File: Calibration/HcalCalibAlgos/test/HcalIsoTrkAnalyzer.cc [hunk 5]
```diff
@@ -396,21 +449,75 @@ void HcalIsoTrkAnalyzer::analyze(edm::Event const& iEvent, edm::EventSetup const
...
+    for (unsigned int iHLT=0; iHLT<triggerResults->size(); iHLT++) {
+      int hlt    = triggerResults->accept(iHLT);
+      if (trigNames_.size() > 0) {
```
Comment (davidlange6) [inline Calibration/HcalCalibAlgos/test/HcalIsoTrkAnalyzer.cc]: this check should go outside the loops

File: Calibration/HcalCalibAlgos/test/HcalIsoTrkAnalyzer.cc [hunk 6]
```diff
@@ -928,10 +1032,69 @@ double HcalIsoTrkAnalyzer::rhoh(const edm::Handle<CaloTowerCollection>& tower) {
...
+				     std::vector<double> *hitEnergies) {
+  double ehcal(0);
+  if (unCorrect_) {
+    for (unsigned int k=0; k<ids.size(); ++k) {
+      double corr = (respCorrs->getValues(ids[k]))->getValue();
+      if (corr != 0) edet[k] /= corr;
+      ehcal += edet[k];
+    }
+  } else {
+    for (auto en : edet) ehcal += en;
```
Comment (davidlange6) [inline Calibration/HcalCalibAlgos/test/HcalIsoTrkAnalyzer.cc]: const& en

File: Calibration/HcalCalibAlgos/test/HcalIsoTrkAnalyzer.cc [hunk 7]
```diff
@@ -928,10 +1032,69 @@ double HcalIsoTrkAnalyzer::rhoh(const edm::Handle<CaloTowerCollection>& tower) {
...
+      HcalDetId id = hdc_->mergedDepthDetId(HcalDetId(ids[k]));
+      auto itr = hitMap.find(id);
+      if (itr == hitMap.end()) {
+	hitMap[id] = edet[k];
+      } else {
+	(itr->second) += edet[k];
+      }
+    }
+    for (auto hit : hitMap) {
+      detIds->emplace_back(hit.first.rawId());
```
Comment (davidlange6) [inline Calibration/HcalCalibAlgos/test/HcalIsoTrkAnalyzer.cc]: add reserve calls for detIds and hitEnergies

**Comment by bsunanda** [comment]
@cmsbuild Please test

File: Calibration/HcalCalibAlgos/test/HcalIsoTrkAnalyzer.cc (hunk 1 above)
Comment (bsunanda) [reply Calibration/HcalCalibAlgos/test/HcalIsoTrkAnalyzer.cc]: done

File: Calibration/HcalCalibAlgos/test/HcalIsoTrkAnalyzer.cc (hunk 2 above)
Comment (bsunanda) [reply Calibration/HcalCalibAlgos/test/HcalIsoTrkAnalyzer.cc]: done

File: Calibration/HcalCalibAlgos/test/HcalIsoTrkAnalyzer.cc (hunk 3 above)
Comment (bsunanda) [reply Calibration/HcalCalibAlgos/test/HcalIsoTrkAnalyzer.cc]: done

File: Calibration/HcalCalibAlgos/test/HcalIsoTrkAnalyzer.cc (hunk 4 above)
Comment (bsunanda) [reply Calibration/HcalCalibAlgos/test/HcalIsoTrkAnalyzer.cc]: done

File: Calibration/HcalCalibAlgos/test/HcalIsoTrkAnalyzer.cc (hunk 5 above)
Comment (bsunanda) [reply Calibration/HcalCalibAlgos/test/HcalIsoTrkAnalyzer.cc]: done

File: Calibration/HcalCalibAlgos/test/HcalIsoTrkAnalyzer.cc (hunk 7 above)
Comment (bsunanda) [reply Calibration/HcalCalibAlgos/test/HcalIsoTrkAnalyzer.cc]: done

**Comment by bsunanda** [comment]
@arunhep @lpernie Please see and approve this PR

**Comment by arunhep** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Pull request merged**